hc_core/src/instance.rs

```diff
@@ -1,46 +1,146 @@
-use error::HolochainError;
+//use error::HolochainError;
 use state::*;
-use std::collections::VecDeque;
+use std::sync::mpsc::*;
+use std::sync::{Arc, RwLock, RwLockReadGuard};
+use std::thread;
+use std::time::Duration;
 
-#[derive(Clone)]
+//#[derive(Clone)]
 pub struct Instance {
-    state: State,
-    pending_actions: VecDeque<Action>,
+    state: Arc<RwLock<State>>,
+    action_channel: Sender<ActionWrapper>,
+    observer_channel: Sender<Observer>,
 }
 
+type ClosureType = Box<FnMut(&State) -> bool + Send>;
+
+pub struct Observer {
+    sensor: ClosureType,
+    done: bool,
+}
+
+impl Observer {
+    fn check(&mut self, state: &State) {
+        self.done = (self.sensor)(state);
+    }
+}
+
+static DISPATCH_WITHOUT_CHANNELS: &str = "dispatch called without channels open";
+
 impl Instance {
-    pub fn dispatch(&mut self, action: Action) {
-        self.pending_actions.push_back(action);
+    pub fn dispatch(&mut self, action: Action) -> ActionWrapper {
+        let wrapper = ActionWrapper::new(action);
+        self.action_channel
+            .send(wrapper.clone())
+            .unwrap_or_else(|_| panic!(DISPATCH_WITHOUT_CHANNELS));
+        wrapper
+    }
+
+    pub fn dispatch_and_wait(&mut self, action: Action) {
+        let wrapper = ActionWrapper::new(action);
+        let wrapper_clone = wrapper.clone();
+
+        let (sender, receiver) = channel::<bool>();
+        let closure = move |state: &State| {
+            if state.history.contains(&wrapper_clone) {
+                sender
+                    .send(true)
+                    .unwrap_or_else(|_| panic!(DISPATCH_WITHOUT_CHANNELS));
```

Comment:
use .expect() instead

Reply:
heh!  That is how we had it to start with, but clippy throws a warning about that and gives you this line instead.  I don't know why...

Reply:
...